src/rgw/rgw_data_sync.cc

```diff
@@ -1315,15 +1315,26 @@ class RGWDataSyncSingleEntryCR : public RGWCoroutine {
     reenter(this) {
       if (state->obligation) {
         // this is already syncing in another DataSyncSingleEntryCR
-        if (state->obligation->timestamp < obligation.timestamp) {
-          // cancel existing obligation and overwrite it
-          tn->log(10, SSTR("canceling existing obligation " << *state->obligation));
-          complete = std::move(*state->obligation);
-          *state->obligation = std::move(obligation);
-          state->counter++;
+        if (!state->obligation->gen || !obligation.gen || 
+            (state->obligation->gen >= obligation.gen)) {
+          // one if the generations is missing, or new obligation has older/same generation
+          // compare only based on time
+          if (state->obligation->timestamp < obligation.timestamp) {
+            // cancel existing obligation and overwrite it
+            tn->log(10, SSTR("canceling existing obligation " << *state->obligation));
+            complete = std::move(*state->obligation);
+            *state->obligation = std::move(obligation);
```

Comment:
if the generations don't match here, this overwrite isn't going to do the right thing

`state->obligation` is tracking another call to `DataSyncSingleEntryCR` that's still inside of `RGWRunBucketSourcesSyncCR`. if we overwrite that here, then that racing call to `DataSyncSingleEntryCR` will complete _our_ obligation instead of its own. if `RGWRunBucketSourcesSyncCR` fails with an error, it would write _our_ generation number to the error repo, losing track of the original obligation

when generations *do* match, this overwrite makes sense. if we get a new obligation to sync up to time T2, we can throw away an earlier obligation at T1 without losing any information - syncing up to T2 implies that we sunc up to T1

but bilog shards of different generations are different things entirely. an obligation to sync gen=2 does not imply an obligation to sync gen=1 - the sync of gen=2 *depends on* the obligation to finish sync of gen=1